### PR TITLE
Add unused variables linter

### PR DESCRIPTION
**Pull request by larsk21**
Add a linter for reporting unused variables in Lean code using Lean's linter API. A variable is reported as unused if it has a declaration, but no uses, except when
- its name starts with an underscore
- it is defined with the `variable` command
- it appears in the signature of a constructor or structure field
- it appears in the signature of a constant without a definition or an axiom
- it appears in the signature of a function that is defined elsewhere (`extern`, `implementedBy`)

### CODE REVIEW

**Comment by larsk21** [comment]
### Examples of unused variables in this repo

*Whether all of these should be reported as unused is open for debate.*

unnecessary variables:
```lean
-- src/Lean/Elab/PreDefinition/Structural/IndPred.lean
def mkIndPredBRecOn (recFnName : Name) (recArgInfo : RecArgInfo) (value : Expr) : M Expr := do
  ...
  forallBoundedTelescope FType (some 1) fun below _ => do
    let main ← mkFreshExprSyntheticOpaqueMVar FType
     -- ^
    let below := below[0]
    let valueNew     ← replaceIndPredRecApps recFnName recArgInfo motive value
    let Farg         ← mkLambdaFVars (recArgInfo.indIndices ++ #[major, below] ++ otherArgs) valueNew
    let brecOn       := mkApp brecOn Farg
    return mkAppN brecOn otherArgs
```

unused pattern variables:
```lean
-- src/Lean/Data/Name.lean
def updatePrefix : Name → Name → Name
  | anonymous, newP => anonymous
            -- ^
  | str p s _, newP => Name.mkStr newP s
     -- ^
  | num p s _, newP => Name.mkNum newP s
     -- ^
```

unused loop variables:
```lean
-- src/Lean/Elab/BuiltinCommand.lean
private def popScopes (numScopes : Nat) : CommandElabM Unit :=
  for i in [0:numScopes] do
   -- ^
    popScope
```

unused function arguments:
```lean
-- src/Lean/Elab/Tactic/Simp.lean
def dsimpLocation (ctx : Simp.Context) (fvarIdToSimp : FVarIdToLemmaId := {}) (loc : Location) : TacticM Unit := do
  ...
where
  go (fvarIdsToSimp : Array FVarId) (simplifyTarget : Bool) (fvarIdToSimp : Lean.Meta.FVarIdToLemmaId) : TacticM Unit := do
                                                          -- ^
    let mvarId ← getMainGoal
    let result? ← dsimpGoal mvarId ctx (simplifyTarget := simplifyTarget) (fvarIdsToSimp := fvarIdsToSimp)
    match result? with
    | none => replaceMainGoal []
    | some mvarId => replaceMainGoal [mvarId]
```

variables that were meant to be used:
```lean
-- src/Lean/Data/Json/Basic.lean
def setObjVal! : Json → String → Json → Json
  | obj kvs, k, v => obj <| kvs.insert compare k v
  | j      , _, _ => panic! "Json.setObjVal!: not an object: {j}"
 -- ^                                                         ^ (missing `s!`)

-- src/Lean/Data/JsonRpc.lean
instance : ToString RequestID where
  toString
  | RequestID.str s => s!"\"s\""
               -- ^         ^ (probably missing `{}`)
  | RequestID.num n => toString n
  | RequestID.null => "null"
```

variables that are used in a structure constructor, but actually unnecessary:
```lean
-- src/Lean/Compiler/IR/Sorry.lean
def updateSorryDep (decls : Array Decl) : CompilerM (Array Decl) := do
  let (_, s) ← Sorry.collect decls |>.run {}
  return decls.map fun decl =>
    match decl with
    | Decl.fdecl f xs t b info =>
                       -- ^
      match s.localSorryMap.find? f with
      | some g => Decl.fdecl f xs t b { info with sorryDep? := some g }
                                     -- ^ (`sorryDep?` is the only field of the structure)
      | _ => decl
    | _ => decl
```

variables that are used as documentation in dependent arrows:
```lean
-- src/Lean/Elab/PreDefinition/WF/Fix.lean
private partial def processSumCasesOn (x F val : Expr) (k : (x : Expr) → (F : Expr) → (val : Expr) → TermElabM Expr) : TermElabM Expr := do
                                                          -- ^            ^            ^
  ...
```

variables that are used as documentation in `Syntax` patterns (`$_` is not allowed):
```lean
-- src/Lean/Elab/Match.lean
private def getMatchGeneralizing? : Syntax → Option Bool
  | `(match (generalizing := true)  $[$motive]? $discrs,* with $alts:matchAlt*) => some true
                                    -- ^         ^              ^
  | `(match (generalizing := false) $[$motive]? $discrs,* with $alts:matchAlt*) => some false
                                    -- ^         ^              ^
  | _ => none
```

File: src/Lean/Linter/Util.lean [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+def publishMessage
+  (content : String) (range : String.Range) (severity : MessageSeverity := .information) : CommandElabM Unit :=
+do
+  let ctx := (← read)
+  let messages := (← get).messages |>.add (mkMessageCore ctx.fileName ctx.fileMap content severity range.start range.stop)
+  modify ({ · with messages := messages })
+
+abbrev SyntaxStack := List (Syntax × Nat)
+
+partial def anyWithStack (pred : Syntax → SyntaxStack → Bool) (stx : Syntax) : Bool :=
```
Comment (Kha) [inline src/Lean/Linter/Util.lean:23]: If we assume that we will only need this to find nested `Syntax` by its lexical range, it might be easier to have something like
```
findSyntaxStack? (root child : Syntax) : Option SyntaxStack
```
where for performance we can skip recursing into nodes that do not contain `child.getRange`

File: src/Lean/Linter/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+      stx.getRange?.isEqSome range &&
+      stx.isOfKind decl.stx.getKind && (
+        matchesUnusedPattern stx stack ||
+        isTopLevelDecl constDecls stx stack ||
+        isVariable stx stack ||
+        isInCtor stx stack ||
+        isInStructBinder stx stack ||
+        isInConstantWithoutDef stx stack ||
+        isInAxiom stx stack ||
+        isInDefWithForeignDefinition stx stack
```
Comment (Kha) [inline src/Lean/Linter/Basic.lean:64]: Now how to make this extensible... :)

**Review by Kha — COMMENTED** [review]
Very nice. Apart from the dependent arrows that we discussed before, to me all the examples are either clear wins or very reasonable to change.

File: src/util/shell.cpp [hunk 3]
```diff
@@ -542,6 +542,7 @@ extern "C" LEAN_EXPORT int lean_main(int argc, char ** argv) {
                 try {
                     check_optarg("D");
                     opts = set_config_option(opts, optarg);
+                    forwarded_args.push_back(string_ref("-D" + std::string(optarg)));
```
Comment (Kha) [inline src/util/shell.cpp:545]: Could you move this to a separate commit?

File: src/Lean/Linter/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    )) then continue
+
+    if uses.isEmpty then
+      let pos := fileMap.toPosition range.start
+      publishMessage s!"unused variable `{localDecl.userName}` at {pos.line}:{pos.column + 1}" range
```
[the remark this answers is not in the document]
Comment (larsk21) [reply src/Lean/Linter/Basic.lean:69]: Yeah, the tests were the only reason I included the position in the message.

The linter messages have `MessageSeverity.information`, which according to `Lean.Message.toString` means that the position is not printed. The `lean` tests  do enable `printMessageEndPos`, but this only adds the `endPos` in addition to the usual position to messages with severity `warning` or `error`.

https://github.com/leanprover/lean4/blob/56cd6c1ff519d8be39f35875b5bf8719e6101c20/src/Lean/Message.lean#L199-L210

File: src/Lean/Linter/Basic.lean [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+def unusedVariables : Linter := fun stx => do
+  let some stxRange := stx.getRange?
+    | pure ()
+
+  let fileMap := (← read).fileMap
+  let refs := dedupReferences <| combineFvars <| findReferences fileMap (← get).infoState.trees.toArray
+
+  let mut vars : HashMap FVarId (Option Reference × Array Reference) := .empty
```
[the remark this answers is not in the document]
Comment (larsk21) [reply src/Lean/Linter/Basic.lean:23]: It's basically the same logic as in `RefInfo.addRef` and `ModuleRefs.addRef` combined with the splitting in `fvar`s and `const`s, but reimplementing this allows to keep the additional information in `Reference`. If we wanted to reuse `ModuleRefs`, we would need to include the additional information in `RefInfo`, which for now only includes ranges. This would make the `ToJson` instance a little more complicated, but I don't know how we would handle the `FromJson` conversion.

File: src/Lean/Linter/Basic.lean (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kha) [reply src/Lean/Linter/Basic.lean:74]: Right, thanks. That does pose the question of whether (most) lints should be `information` or `warning` (and we might want an option for raising them to `error` in either case). I think most compilers/linters style them as warnings, which would also help to distinguish them from the existing set of informational messages that are usually of a more temporary nature.

[the remark this answers is not in the document]
Comment (Kha) [reply src/Lean/Linter/Basic.lean:74]: > (and we might want an option for raising them to `error` in either case)

Actually, if we do make them warnings, a `-Werror`-style option that raises *all* warnings to errors would make more sense.

**Comment by Kha** [comment]
Leo pointed out the following special case to me, where at least currently there is no connection between the different `h₂`s in the info tree:
```
theorem foo (h₁ : x = 0) (h₂ : y = x) : y = 0 := by
  subst h₁
  -- We now have `h₂ : y = 0`
  exact h₂
```
This is probably an issue for quite a few (and potentially user-written) tactics, so I'm wondering if we should look for references on the `Expr` level after all. Perhaps in combination with the `References` module so as to avoid edge cases like
```
let mut x := 0  -- technically dead, but perhaps not helpful as an "unused variable" lint
if b then x := 1 else x := 2
```

In case we expect that there may be further issues in practice, there is also the option of merging the lint as opt-in for now so that people can test-drive it (which I assume can be done very easily with Lake's `moreLean/ServerArgs`) and we can further improve it without annoying too many people at the same time. In that case I'm fine with merging it as is.

**Comment by gebner** [comment]
> This is probably an issue for quite a few (and potentially user-written) tactics, so I'm wondering if we should look for references on the `Expr` level after all.

Oh I didn't realize this didn't even look at the expression.  Then there's of course lots of obvious corner cases, here are a few off the top of my head that I haven't seen mentioned so far:
```lean
def f (x : Nat) := ‹_›  -- false positive
def g [Inhabited α] := 42  -- false negative
def h :=
  let inst : Inhabited Nat := ⟨42⟩ -- false positive
  (default : Nat)
```

File: src/Lean/Linter/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    stx.getId.toString.startsWith "_"
+  isVariable (_ : Syntax) (stack : SyntaxStack) :=
+    stackMatches stack [`null, none, `null, `Lean.Parser.Command.variable]
+  isInCtor (_ : Syntax) (stack : SyntaxStack) :=
+    stackMatches stack [`null, none, `null, `Lean.Parser.Command.optDeclSig, `Lean.Parser.Command.ctor]
+  isInStructBinder (_ : Syntax) (stack : SyntaxStack) :=
+    stackMatches stack [`null, none, `null, `Lean.Parser.Command.optDeclSig, `Lean.Parser.Command.structSimpleBinder]
+  isInConstantWithoutDef (_ : Syntax) (stack : SyntaxStack) :=
+    stackMatches stack [`null, none, `null, `Lean.Parser.Command.declSig, `Lean.Parser.Command.constant] &&
+    (stack.get? 4 |>.any fun (stx, _) => stx[3].getNumArgs == 0)
```
[the remark this answers is not in the document]
Comment (larsk21) [reply src/Lean/Linter/Basic.lean:93]: But is there a benefit of writing the names of the parameters in this case? Or is there a case where we cannot write an underscore instead of an identifier?

**Comment by larsk21** [comment]
> Oh I didn't realize this didn't even look at the expression. Then there's of course lots of obvious corner cases, here are a few off the top of my head that I haven't seen mentioned so far:
> 
> ```lean
> def f (x : Nat) := ‹_›  -- false positive
> def g [Inhabited α] := 42  -- false negative
> def h :=
>   let inst : Inhabited Nat := ⟨42⟩ -- false positive
>   (default : Nat)
> ```

I agree that the `α` should be marked as an unused variable. For the other two, I'm unsure if these are false positives, since we could instead write:
```lean
def f (_ : Nat) := ‹_›
def h :=
  have : Inhabited Nat := ⟨42⟩
  (default : Nat)
```
The first example is similar to this test where `HQ` is also marked as unused. Currently the *unused* marker doesn't always mean that the value of the variable is not needed, but that variable is never referenced by name.
https://github.com/leanprover/lean4/blob/f8b1b210545e39a3d8f037316f0f3a08f1aee992/tests/lean/linterUnusedVariables.lean#L8-L11

**Comment by gebner** [comment]
> [...] we could instead write:

```lean
def f (_ : Nat) := ‹_›
```

This is a semantic change, with the previous version you could write `f (x := 42)`, with this one you can't.  Argument names are, for better or for worse, part of the public API.

```lean
def h :=
  have : Inhabited Nat := ⟨42⟩
  (default : Nat)
```
Usually you need `let` instead of `have` for local instances.  Unfortunately `let : Inhabited Nat := ⟨42⟩` doesn't work, but that's fixable.

**Comment by larsk21** [comment]
> Argument names are, for better or for worse, part of the public API.
I agree. As it is now, the linter reports function arguments as unused like any other variable - maybe I should change this.

**Comment by Kha** [comment]
Right, there are two slightly different potential lints: a) "this binder can be safely removed (it is unused)" and b) "this binder can be safely made anonymous (its *name* is unused)". b) is, as discussed, problematic in public signatures, and not really solvable for some tactics. So we should probably focus on a) using a combination of lexical `Expr`-level references.

File: src/Lean/Linter/Basic.lean (hunk 6 above)
[the remark this answers is not in the document]
Comment (Kha) [reply src/Lean/Linter/Basic.lean:93]: As in the more recent discussion below, we shouldn't rename parameters. Essentially all the arguments for keeping parameter names of definition-less `constant`s apply to those with definitions as well.

**Comment by Kha** [comment]
> I'm wondering if we should look for references on the `Expr` level after all.

Easier said than done, of course. There is no direct association between the free variables from the info tree and the bound variables in the final term. However, since the above issues are mostly about tactics, it might be sufficient to go through all the terms assigned to `TacticInfo` mvars. We should make sure not to visit any mvar twice though to avoid quadratic overhead.

That would leave the instance binders. I don't immediately see a better solution than just whitelisting them.

**Comment by Kha** [comment]
Excellent work! Now I really think we should just merge it and then await feedback by users. Even for the many warnings in the stdlib, I can't yet tell if I would be annoyed enough by them that we should turn them off there, so let's just try and see I'd suggest.

If they do tend to be a bit too much in e.g. the stdlib (until we can, say, fix them automatically), we could look into suppressing them in patterns only, which seems to be the vast majority of cases.

File: src/Lean/Linter/Basic.lean [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+              assignments
+            else
+              assignments.insert mvar expr
+        | _ =>
+          assignments)
+
+  let tacticFVarUses : HashSet FVarId ←
+    tacticMVarAssignments.foldM (init := .empty) fun uses _ expr => do
+      let (_, s) ← StateT.run (s := uses) <| expr.forEach' (fun expr => do
+        match expr with
```
Comment (Kha) [inline src/Lean/Linter/Basic.lean:56]: Note that this will not visit assignments of nested metavariables, so we could miss a part of the proof. But since we do this for every tactic step, I don't think this is an issue for the current set of tactics.

**Comment by gebner** [comment]
Is there something like the `@[nolint unused_arguments]` we have in mathlib to disable the lint?  There's 84 definitions in mathlib where one of the arguments is *intentionally* unused.

**Comment by gebner** [comment]
> I see that you ported the mathlib 3 infrastructure using an attribute

I didn't want to make any big refactorings yet.  Most of the mathlib linters are 1) per declaration and 2) non-local.  An attribute makes perfect sense here, since we need to run the linters with all of mathlib imported (when an option would have long been discarded).

> but using options has the distinct advantage that we can apply them at many different levels:

That this is impossible is a feature, not a bug. Opting out of best practices should be annoying, and not happen accidentally.  We could of course add a "no nolint on sections/modules" linter. :smile: 

> Then there is probably a larger conversation to be had about how the Lean and mathlib approaches to linters should be correlated/integrated/coexist, but maybe not in this PR.

Indeed, it would be great if e.g. disabling a linter had the same syntax in both cases.  But I don't see either of the two APIs subsume the other.  Non-local checks (like checking for simp-normal form, or for typeclass diamonds) clearly don't fit into Lean concept of "lint the command syntax right here and now".  Syntactical checks (like for line-length, comments, deprecated syntax, etc.) clearly don't fit into the mathlib concept of "check the full mathlib environment".

Even the unused variable linter doesn't neatly fit into either category.  You need to look at the AST to flag `by have x := 42; exact 1`※.  But on the other hand we also want to check declarations generated by metaprograms, which do not have an input syntax.

※ I just noticed that `have` behaves differently in Lean 4 than it did in Lean 3.  In Lean 3 the have disappeared, so you could use it to declare local instances without affecting the elaborated term.  But now it produces `let_fun`.

**Comment by larsk21** [comment]
With these fine-grained options to disable linters, do we still want the unused variable pattern `_x`, i.e. variables starting with an underscore are ignored by the linter?

File: src/Lean/Linter/Basic.lean (hunk 7 above)
Comment (Kha) [reply src/Lean/Linter/Basic.lean:66]: (nevermind, you're iterating over `eAssignments` directly, not `goalsAfter`)

**Comment by Kha** [comment]
It's still useful for other kinds of variables, no? And I assume most code will (eventually) not make use of these options.

**Comment by larsk21** [comment]
Sure, it can be used on every variable detected by the linter.

**Comment by larsk21** [comment]
Some statistics from a `stage2` build of the Lean repo:

| category | number of unused variables | percentage of all unused variables |
|-|-:|-:|
| all | 1337 | 100% |
| function arguments | 353 | 26.4% |
| pattern variables | 571 | 42.7% |

**Comment by Kha** [comment]
Interesting, I would have thought it would be more pattern variables. In that case I'm fine with merging it as is, and I believe @leodemoura as well. @gebner Do you have anything in mind that should be addressed right now?

> Opting out of best practices should be annoying, and not happen accidentally.

I don't feel that strongly about opting out btw. Clippy/Rust also supports opt-outs at the crate & module level. And it's still not really accidental, is it?

**Comment by leodemoura** [comment]
> Interesting, I would have thought it would be more pattern variables. In that case I'm fine with merging it as is, and I believe @leodemoura as well. @gebner Do you have anything in mind that should be addressed right now?

Yes, I am fine with merging as is. I will be happy to address the unused variable warnings.
